rest_framework/serializers.py

```diff
@@ -1242,9 +1244,10 @@ def build_standard_field(self, field_name, model_field):
             # `allow_blank` is only valid for textual fields.
             field_kwargs.pop('allow_blank', None)
 
-        if postgres_fields and isinstance(model_field, postgres_fields.JSONField):
+        if (postgres_fields and isinstance(model_field, postgres_fields.JSONField)) or \
+                (hasattr(models, 'JSONField') and isinstance(model_field, models.JSONField)):
             # Populate the `encoder` argument of `JSONField` instances generated
-            # for the PostgreSQL specific `JSONField`.
+            # for the model `JSONField`.
             field_kwargs['encoder'] = getattr(model_field, 'encoder', None)
```

Comment:
django 3.1 JSONField adds `decoder` attr
this could be
```python
is_django_jsonfield = hasattr(models, 'JSONField') and isinstance(model_field, models.JSONField)
if (postgres_fields and isinstance(model_field, postgres_fields.JSONField)) or is_django_jsonfield:
    # Populate the `encoder` argument of `JSONField` instances generated
    # for the model `JSONField`.
    field_kwargs['encoder'] = getattr(model_field, 'encoder', None)
    if is_django_jsonfield:
        field_kwargs['decoder'] = getattr(model_field, 'decoder', None)
```

Reply:
The serializer JSONField only supports `encoder`, though, so that would also require adding `decoder` support to the serializer field. Since it currently uses `encoder` for both, that could potentially be a breaking change, and seems outside the scope of just fixing Django 3.1 support.

Reply:
> would also require adding decoder `support` to the serializer field
Yes, I think it should be added (and tests)

> Since it currently uses encoder for both, that could potentially be a breaking change
No, it's not used (https://github.com/encode/django-rest-framework/blob/master/rest_framework/fields.py#L1780 

```python
            if self.binary or getattr(data, 'is_json_string', False):
                if isinstance(data, bytes):
                    data = data.decode()
                return json.loads(data)
```

I think it should be updated as

```diff
diff --git a/rest_framework/fields.py b/rest_framework/fields.py
index da2dd54b..f218713f 100644
--- a/rest_framework/fields.py
+++ b/rest_framework/fields.py
@@ -1758,6 +1758,7 @@ class JSONField(Field):
     def __init__(self, *args, **kwargs):
         self.binary = kwargs.pop('binary', False)
         self.encoder = kwargs.pop('encoder', None)
+        self.decoder = kwargs.pop('decoder', None)
         super().__init__(*args, **kwargs)
 
     def get_value(self, dictionary):
@@ -1777,7 +1778,7 @@ class JSONField(Field):
             if self.binary or getattr(data, 'is_json_string', False):
                 if isinstance(data, bytes):
                     data = data.decode()
-                return json.loads(data)
+                return json.loads(data, cls=self.decoder)
             else:
                 json.dumps(data, cls=self.encoder)
         except (TypeError, ValueError):
```

(I don't think it is worth check for django/jsonfield version and always allow `decoder` flag)

Reply:
> > Since it currently uses encoder for both, that could potentially be a breaking change
> 
> No, it's not used (https://github.com/encode/django-rest-framework/blob/master/rest_framework/fields.py#L1780
> 
> ```python
>             if self.binary or getattr(data, 'is_json_string', False):
>                 if isinstance(data, bytes):
>                     data = data.decode()
>                 return json.loads(data)
> ```
> 

Ok, I misread that, I just saw it using encoder right below that:

https://github.com/encode/django-rest-framework/blob/aed74961ba03e3e6f53c468353f4e255eb788555/rest_framework/fields.py#L1780-L1782

> > would also require adding `decoder` support to the serializer field
> 
> Yes, I think it should be added (and tests)
> 

I'll add `decoder`, then.